File: java/src/main/java/com/test/java/animals/Animal.java
package com.test.java.animals;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.CUSTOM, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = Cat.class, name = "cat"),
        @JsonSubTypes.Type(value = Dog.class, name = "dog")
})
public interface Animal {
    String getNoise();
}

File: kotlin/src/main/kotlin/com/test/kotlin/animals/Animal.kt
package com.test.kotlin.animals

import com.fasterxml.jackson.annotation.JsonTypeInfo

@JsonTypeInfo(use = JsonTypeInfo.Id.CUSTOM, property = "type")
//@JsonSubTypes(JsonSubTypes.Type(value = Cat::class, name = "cat"), JsonSubTypes.Type(value = Dog::class, name = "dog"))
interface Animal {
    val noise: String
}
